src/ai/backend/agent/resources.py

```diff
@@ -523,6 +524,61 @@ async def _wrap_future(fut: asyncio.Future) -> None:
     return slot_allocs
 
 
+async def scan_gpu_alloc_map(
+    kernel_ids: Sequence[KernelId], scratch_root: Path, concurrency: Optional[int] = None
+) -> dict[DeviceId, Decimal]:
+    """
+    Fetch the current allocated amounts for fractional gpu from
+    ``/home/config/resource.txt`` files in the kernel containers managed by this agent.
+    """
+    if concurrency is None:
+        concurrency = os.cpu_count() or 4
+
+    semaphore = asyncio.Semaphore(concurrency)
+
+    async def _read_kernel_resource_spec(path: Path) -> dict[DeviceId, Decimal]:
+        alloc_map: dict[DeviceId, Decimal] = defaultdict(lambda: Decimal(0))
+
+        async with semaphore:
+            try:
+                loop = asyncio.get_running_loop()
+                content = await loop.run_in_executor(None, path.read_text)
+                resource_spec = KernelResourceSpec.read_from_string(content)
+
+                if cuda := resource_spec.allocations.get(DeviceName("cuda")):
+                    if cuda_shares := cuda.get(SlotName("cuda.shares")):
+                        for device_id, shares in cuda_shares.items():
+                            alloc_map[device_id] += Decimal(shares)
+
+                    if cuda_device := cuda.get(SlotName("cuda.device")):
+                        for device_id, device in cuda_device.items():
+                            alloc_map[device_id] += Decimal(device)
```

Comment:
Oh, I'm sorry. I think I misunderstood that last time.
I thought the coroutine was being created by going around the for loop, but it wasn't.
Please remove the semaphore code and I'll approve.

Reply:
Ah... the for loop slipped into the upper layer...

Reply:
Rather than limiting the number of coroutines that actually work with semaphores, I wanted to create fewer coroutines with fewer iterations of the for loop itself.

Reply:
We decided to remove the concurrency restriction (semaphore or batching) here since the number of coroutines is not too much.